Cog-Creators/Red-DiscordBot
# Add support for sdists and git-archive to `_get_version()`

### Description of the changes

Recently (Git 2.35 - January 2022), git-archive's `export-subst` has gotten support for adding the output of `git describe --tags` to it. With that feature, `_get_version()` can provide the same information for git-archive as for full-fledged git repositories. This should allow git-archive to insert the output of `git describe --tags` into the `.git_archive_info.txt`. 

Additionally, I also added support for sdists which ensures that a wheel built from an sdist will keep the sdist's version number.

In addition to [adding support for git-archive](https://github.com/Cog-Creators/Red-DiscordBot/commit/0b6d5408a282d995e621aad62e41934c6112508e) and [sdists](https://github.com/Cog-Creators/Red-DiscordBot/commit/90c839c3b7d98733436775973661cbe79461777c), I also [split out the get version logic per retrieval method](https://github.com/Cog-Creators/Red-DiscordBot/commit/41edc1987756fe43705cfe9647af72c396c4d982) to make the code cleaner.

### How to test?

In all of the below scenarios, you can easily confirm the generated version by running `redbot --version`.

In order to test the support for git-archive, run:
```
python -m pip install https://github.com/jack1142/Red-DiscordBot/archive/support_git_archive.zip
python -m pip install https://github.com/jack1142/Red-DiscordBot/archive/support_git_archive.tar.gz
```

Other scenarios should continue to work as well:
```
python -m pip install -U git+https://github.com/jack1142/Red-DiscordBot@support_git_archive
```
```
python -m pip install -U pip install git+https://github.com/Cog-Creators/Red-DiscordBot@refs/pull/5814/merge
```
```
git clone https://github.com/jack1142/Red-DiscordBot -b support_git_archive
cd Red-DiscordBot
python -m pip install -e .
```

As for testing sdists, you can just run `python -m build` in the root of the repository (you'll need to `pip install build` first) and verify that the built wheel has a proper version number.

### Have the changes in this PR been tested?

Yes
